src/common.c

```diff
@@ -0,0 +1,201 @@
+//
+// Created by alfylinux on 21/10/18.
+//
+
+#include "common.h"
+
+
+int semInit()
+{
+	int semId = semget(IPC_PRIVATE,3, IPC_CREAT | IPC_EXCL | 0666);
+	if(semId == -1)
+	{
+		perror("Create Sem-s take error:");
+		return -1;
+	}
+
+	//enum semName {wantWrite=0,readWorking=1,writeWorking=2}; number is Id of sem
+	unsigned short semStartVal[3] = {0, 0, 1};
+
+	//setup 3 semaphore in system5
+	if(semctl(semId, 0, SETALL, semStartVal))
+	{
+		perror("set Sem take error:");
+		return -1;
+	}
+
+	printf("SEMAFORO Created\n");
+	semInfo(semId);
+
+	return semId;
+}
+
+int lockWriteSem(int semId)
+{
+	struct sembuf sem;
+	sem.sem_flg = SEM_UNDO;
+
+	//increase counter wantWrite
+	sem.sem_num = wantWrite;
+	sem.sem_op = +1;
+	SEM_wantWrite:
+	if(semop(semId, &sem, 1))
+	{
+		switch (errno)
+		{
+			case EINTR:
+				goto SEM_wantWrite;
+				break;
+			default:
+				perror("increase wantWrite semCount error:");
+				return -1;
+				break;
+		}
+	}
+
+	//wait until =0 readerWork
+	sem.sem_num = readWorking;
+	sem.sem_op = 0;
+
+	SEM_waitReaders:
+	if(semop(semId, &sem, 1))
+	{
+		switch (errno)
+		{
+			case EINTR:
+				goto SEM_waitReaders;
+				break;
+			default:
+				perror("wait until 0 readWorking take error:");
+				return -1;
+				break;
+		}
+	}
+	//wait noThread already work
+	sem.sem_num = writeWorking;
+	sem.sem_op = -1;
+
+	SEM_writeWorking:
+	if(semop(semId, &sem, 1))
+	{
+		switch (errno)
+		{
+			case EINTR:
+				goto SEM_writeWorking;
+				break;
+			default:
+				perror("lock writeWorking error:");
+				return -1;
+				break;
+		}
+	}
+	return 0;
+}
+
+int unlockWriteSem(int semId)
+{
+	struct sembuf sem;
+	sem.sem_flg = SEM_UNDO;
+
+	//signal finish writing work
+	sem.sem_num = writeWorking;
+	sem.sem_op = 1;
+
+	SEM_writeWorking:
+	if(semop(semId, &sem, 1))
+	{
+		switch (errno)
+		{
+			case EINTR:
+				goto SEM_writeWorking;
+				break;
+			default:
+				perror("unlock writeWorking error:");
+				return -1;
+				break;
+		}
+	}
+	//reduce counter of wantWrite
+	sem.sem_num = wantWrite;
+	sem.sem_op = -1;
+
+	SEM_wantWrite:
+	if(semop(semId, &sem, 1))
```

Comment:
SPACE.

Reply:
it is to highlight the goto before if, you prefer remove it?

Reply:
if(semop(semId, &sem, 1)) -> if (semop(semId, &sem, 1))

Reply:
oooo.... sorry hahaha
now done